design/crd-status.md

```diff
@@ -0,0 +1,237 @@
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: ConfigurationStatus
+metadata:
+  name: config-status
+  namespace: metallb-system
+status:
+    validConfig: false
+    error: "error message here"
+```
+
+**Note:** given the fact that the configuration is parsed both by the speakers and
+by the controller, we might want to expose the status of each component. However,
+there is currently no logic in the configuration parsing that depends on the component.
+For this reason, we can have an initial status that is produced only by the controller
+(which runs in single instance) to validate the status.
+
+As an alternative, we might consider having a state per component, named after the
+single component that produces the status, but in general having a single place to
+check seems more straightforward.
+
+If we go with the _per component_ scenario, we might add a `loadedConfiguration` field
+that exposes the latest loaded configuration by that component. This can't be done
+if we let the controller to expose the single configuration because what's loaded might
+depend on the order the CRs are received with.
+
+#### BGP / BFD Session state
+
+Because of the scalability concerns expressed above, the idea is to produce an instance
+of the resource _per peer / node_, which exposes the state of the session between
+the speaker running on a given node and a given peer.
+
+The name of a given instance will be like `nodename-peer`, and each instance will
+be labeled with the name of the node and the peer it refers to, to make it easier
+to list the status of all the sessions related to a given `BGPPeer` and a given
+node.
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: BGPSessionStatus
+metadata:
+  name: worker0-peer1
+  namespace: metallb-system
+  labels:
+    metallb.io/node: worker0
+    metallb.io/peer: peer1
+status:
+    bgpStatus: Established
+    bfdStatus: Up
+```
+The string exposed is taken directly from the output of FRR.
+
+If BFD is not configured for a given `BGPPeer`, the exposed bfdStatus will be "N/A".
+
+#### Service Announcement - BGP
+
+Given a service and a node, we want to expose the `BGPpeer`s it is advertised to.
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: BGPAdvertisementStatus
+metadata:
+  name: service1-worker0
+  namespace: metallb-system
+  labels:
+    metallb.io/node: worker0
+    metallb.io/service: service1
+status:
+    bgpPeers:
+        - peerA
+        - peerB
+```
+
+The labels will allow to easily discover which services are advertised from
+a given node, or all the peers a given service is advertised to.
+
+#### Service Announcement - L2
+
+The useful information related to L2 are related to the node that is exposing
+the service, and via which interfaces.
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: L2AdvertisementStatus
+metadata:
+  name: service1-worker0
+  namespace: metallb-system
+  labels:
+    metallb.io/node: worker0
+    metallb.io/service: service1
+status:
+    node: worker0
```

Comment:
mostly because without interfaces selector the resource would be empty, and because I thought in L2 the focus is more on the node (we have only one leader here).
The labels are both for discovery (find all the services announced via L2 on this node) and for exposing the status.
This makes me think that we can even get rid of the node in the name.

Reply:
+1 for ServiceL2Status (or changing the bgp one to BGPAdvertisementStatus)